### PR TITLE
Proposal: fastapi-users based authentication backend

### PR DESCRIPTION
**Pull request by dionysio**
Hi,

I really like your repo and I wanted to use it in my new project. I was using fastapi-users to handle most of my auth-related stuff, but I couldn't find an integration between it and your project, so I tried writing my own AuthenticationBackend using fastapi-users tools.

I wanted to check with you if this backend is something you wanted to include in your repo. My suggestion would be to offer it as a variant (I mean one of those optional dependencies like `sqladmin[fastapi-users]`). If you are interested, I can write up the tests and finish it up, if not, no hard feelings!

Thank you!

### CODE REVIEW

**Comment by ix-56h** [comment]
This feature is needed, there is one discussion related to this waiting for answer (https://github.com/aminalaee/sqladmin/discussions/159#discussioncomment-3077027) and one on fastapi-users repo (https://github.com/fastapi-users/fastapi-users/discussions/1218).

I tried to use your commited code, but it seem that it doesn't work for me and i struggle to understand why : 

Fastapi-users definitions :
```python
...
bearer_transport = BearerTransport(tokenUrl="auth/jwt/login")

def get_jwt_strategy() -> JWTStrategy:
    return JWTStrategy(secret=SECRET, lifetime_seconds=3600)

auth_backend = AuthenticationBackend(
    name="jwt",
    transport=bearer_transport,
    get_strategy=get_jwt_strategy,
)

async def get_async_session() -> AsyncGenerator[AsyncSession, None]:
    async with async_session_maker() as session:
        yield session

async def get_user_db(session: Annotated[async_sessionmaker, Depends(get_async_session)]):
    yield SQLAlchemyUserDatabase(session, users.User, users.OAuthAccount)
    
async def get_user_manager(user_db=Depends(get_user_db)):
    yield UserManager(user_db)

fastapi_users_instance = FastAPIUsers[users.User, uuid.UUID](
    get_user_manager,
    [auth_backend],
)```

SQLadmin initialization :
```python
admin = Admin(
    app,  
    engine,  
    authentication_backend=FastapiUsersAuthenticationBackend(
        secret_key=SECRET,
        # fast-users callables
        get_user_manager=get_user_manager,
        get_strategy=get_jwt_strategy)
)
```
Error :
```
app-backend-1  |   File "/usr/local/lib/python3.11/site-packages/sqladmin/application.py", line 604, in login
app-backend-1  |     ok = await self.authentication_backend.login(request)
app-backend-1  |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
app-backend-1  |   File "/app/authentication_fastapi_users.py", line 114, in login
app-backend-1  |     user = await self.get_user_manager().authenticate(credentials)
app-backend-1  |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
app-backend-1  | AttributeError: 'async_generator' object has no attribute 'authenticate'
```

I tried to understand the issue, but after some hours i didn't do any progress.

TLDR: It would be a must have since fastapi-users is pretty popular and a pain in the ass to implement along with sqladmin.

**Comment by regisin** [comment]
@dionysio @ix-56h I'm sorry I'm not sure how to contribute to this PR on github, but I edited the `authentication_fastapi_users.py` file to this to make it work:

```python
import contextlib
from typing import Optional

from fastapi_users.authentication import (
    AuthenticationBackend,
    Transport, Strategy
)
from fastapi import HTTPException, status
from fastapi.security import APIKeyCookie, OAuth2PasswordRequestForm
from fastapi_users.manager import UserManagerDependency
from fastapi_users.router import ErrorCode
from fastapi_users.types import DependencyCallable
from fastapi_users.authentication import Authenticator
from fastapi_users import models
from starlette.requests import Request
from starlette.responses import RedirectResponse
import sqladmin.authentication

from rich import print

class SessionTransport(Transport):
    """
    Modified version of CookieTransport which simply reads cookies and returns tokens instead of full Responses
    """
    scheme: APIKeyCookie

    def __init__(self, name: str = 'fastapiusersauth'):
        super().__init__()
        self.scheme = APIKeyCookie(name=name, auto_error=False)

    async def get_login_response(self, token: str) -> str:
        return token

    async def get_logout_response(self):
        pass

    @staticmethod
    def get_openapi_login_responses_success():
        return {}

    @staticmethod
    def get_openapi_logout_responses_success():
        return {}


class FastapiUsersAuthenticationBackend(sqladmin.authentication.AuthenticationBackend):
    """
        Usage:

        from sqladmin import Admin
        from sqladmin.authentication_fastapi_users import FastapiUsersAuthenticationBackend
        admin = sqladmin.Admin(
            app,  # fastapi app/router
            engine,  # sqla engine
            authentication_backend=FastapiUsersAuthenticationBackend(secret_key='<SUPER SECRET KEY HERE>',
                                                                     # fast-users callables
                                                                     get_user_manager=get_user_manager,
                                                                     get_strategy=get_redis_strategy)
        )
    """
    def __init__(self, secret_key: str,
                 get_user_manager: UserManagerDependency[models.UP, models.ID],
                 get_strategy: DependencyCallable[Strategy[models.UP, models.ID]],
                 get_user_db,
                 get_session,
                 requires_verification: bool = False,
                 name: str = 'admin'):
        super().__init__(secret_key)
        self.get_user_manager = get_user_manager
        self.get_strategy = get_strategy
        self.backend = AuthenticationBackend(
            name=name,
            transport=SessionTransport(),
            get_strategy=get_strategy,
        )
        self.requires_verification = requires_verification


        self.get_session=get_session
        self.get_user_db = get_user_db

        self.get_async_session_context = contextlib.asynccontextmanager(self.get_session)
        self.get_user_db_context = contextlib.asynccontextmanager(self.get_user_db)
        self.get_user_manager_context = contextlib.asynccontextmanager(self.get_user_manager)

    def _redirect_to_login(self, request: Request):
        return RedirectResponse(request.url_for("admin:login"), status_code=302)

    def _validate_user(self, user):
        # code adjusted from fastapi_users.router.auth.login
        if user is None or not user.is_active:
            print(f'not active:{user}')
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=ErrorCode.LOGIN_BAD_CREDENTIALS,
            )
        if self.requires_verification and not user.is_verified:
            print(f'not verified:{user}')
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=ErrorCode.LOGIN_USER_NOT_VERIFIED,
            )
        if not user.is_superuser:
            print(f'not super:{user}')
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail='LOGIN_USER_NOT_SUPERUSER',
            )

    async def authenticate(self, request: Request) -> Optional[RedirectResponse]:
        token = request.session.get("token")
        if not token:
            return False

        async with self.get_async_session_context() as session:
            async with self.get_user_db_context(session) as user_db:
                async with self.get_user_manager_context(user_db) as user_manager:
                    kwargs = {
                        'user_manager': user_manager,
                        f'strategy_{self.backend.name}': self.get_strategy(),
                        self.backend.name: token
                    }
                    authenticator = Authenticator([self.backend], user_manager)
                    user = await authenticator.current_user(optional=True)(**kwargs)
                    print(f'authenticate:user {user}')
                    print(f'authenticate:token {token}')
        try:
            self._validate_user(user)
            return True
        except HTTPException:
            return self._redirect_to_login(request)

    async def login(self, request: Request) -> bool:
        form = await request.form()
        credentials = OAuth2PasswordRequestForm(scope='', username=form.get("username"), password=form.get("password"))

        async with self.get_async_session_context() as session:
            async with self.get_user_db_context(session) as user_db:
                async with self.get_user_manager_context(user_db) as user_manager:
                    user = await user_manager.authenticate(credentials)
                    print(f'login:user {user}')


        self._validate_user(user)

        token = await self.backend.login(self.get_strategy(), user)
        print(f'login:token {token}')
        request.session["token"] = token
        return True

    async def logout(self, request: Request) -> bool:
        request.session.clear()
        return True
```

It looks wonky, but it's working over here.

I basically followed a solution from here: https://github.com/fastapi-users/fastapi-users/discussions/808

and here: https://fastapi-users.github.io/fastapi-users/12.1/cookbook/create-user-programmatically/

**Comment by aminalaee** [comment]
@regisin this is great. You can create a new PR and if possible add some tests for it.

**Comment by dionysio** [comment]
thanks for the updates @regisin! Sorry I went away working on something else and this code was shelved for me.

**Comment by regisin** [comment]
I'm not sure I'm smart enough to do that! Do I contribute with a new PR? Or is it possible to add to this one?

Also, do I need to add `fastapi-users` to `pyproject.toml` somewhere?

**Comment by aminalaee** [comment]
@regisin I can help you. I think its better to start a new PR.

And for the fastapi-users, you need to add it here: https://github.com/aminalaee/sqladmin/blob/0abdef0970693d9ce5bfbdcd4e138c54cda45f44/pyproject.toml#L62

and add some tests in tests/integration

you can also check: https://github.com/aminalaee/sqladmin/blob/main/CONTRIBUTING.md#get-started

**Comment by aminalaee** [comment]
I will start a new PR for this.

**Comment by gpkc** [comment]
@aminalaee Has there been a new PR for this? I can't find it

**Comment by aminalaee** [comment]
No, I haven’t had a chance to work on it. Feel free to submit a PR if you can make it work.
